### PR TITLE
fix: Redirect completed users from onboarding deep link routes to home

### PR DESCRIPTION
**Pull request by MelvinBot**
### Explanation of Change

When a user who has completed onboarding clicks a "Continue Onboarding" deep link (`/onboarding/purpose`) from a Concierge message, nothing happens — the navigation silently fails. This is because the `OnboardingModalNavigator` is conditionally excluded from the navigation stack when `isOnboardingCompleted !== false` (in `AuthScreens.tsx`), so the route can't resolve. The `OnboardingGuard` blindly returns `ALLOW` when onboarding is completed, without checking whether the user is trying to navigate **to** an onboarding route.

This PR adds a check in `OnboardingGuard.evaluate()` that detects when a completed user is navigating to an onboarding route via deep link (RESET action) and redirects them to the home screen instead of silently failing. This follows the same pattern established by `TestDriveModalGuard`, which redirects to `ROUTES.HOME` when a user navigates to a dismissed test drive modal.

### Fixed Issues

$ https://github.com/Expensify/App/issues/79588

### Tests

1. Log into an account that has already completed the onboarding/guided setup flow
2. Navigate to `https://staging.new.expensify.com/onboarding/purpose` directly (or any `/onboarding/*` route)
3. Verify you are redirected to the home/inbox screen instead of nothing happening
4. Try other onboarding routes like `/onboarding/personal-details` and `/onboarding/work-email`
5. Verify all redirect to home
6. Log into a fresh account that has NOT completed onboarding
7. Verify the normal onboarding flow still works correctly
8. Run `npx jest tests/unit/Navigation/guards/OnboardingGuard.test.ts` — all tests pass

- [x] Verify that no errors appear in the JS console

### Offline tests

This change only affects deep link navigation routing logic, which is handled synchronously in the navigation guard. No network requests are involved. Offline behavior is unaffected.

### QA Steps

1. Log into an account that has already completed the onboarding/guided setup flow
2. Navigate to any chat and send message link: `https://staging.new.expensify.com/onboarding/purpose`
3. Verify you are redirected to the home/inbox screen
4. Try other onboarding routes like `/onboarding/personal-details` and `/onboarding/work-email`
5. Verify all redirect to home
6. Log into a fresh account that has NOT completed onboarding
7. Verify the normal onboarding flow still works correctly

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified there are no new alerts related to the `canBeMissing` param for `useOnyx`
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4510fc76bbf5df699a2575bfb49a276af90f3ed7/src/components/LocaleContextProvider.tsx#L80)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4510fc76bbf5df699a2575bfb49a276af90f3ed7/src/components/LocaleContextProvider.tsx#L116-L123)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If new assets were added or existing ones were modified, I verified that:
    - [x] The assets are optimized and compressed (for SVG files, run `npm run compress-svg`)
    - [x] The assets load correctly across all supported platforms.
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

N/A - Navigation guard logic only, no UI changes

</details>

<details>
<summary>Android: mWeb Chrome</summary>

N/A - Navigation guard logic only, no UI changes

</details>

<details>
<summary>iOS: Native</summary>

N/A - Navigation guard logic only, no UI changes

</details>

<details>
<summary>iOS: mWeb Safari</summary>

N/A - Navigation guard logic only, no UI changes

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

N/A - Navigation guard logic only, no UI changes

</details>

### CODE REVIEW

File: src/libs/Navigation/guards/OnboardingGuard.ts [hunk 1]
```diff
@@ -129,6 +130,25 @@ function shouldPreventReset(state: NavigationState, action: NavigationAction) {
     return false;
 }
 
+/**
+ * Check if the navigation action is targeting an onboarding screen.
+ * This handles RESET actions (URL-based navigation / deep links).
+ */
+function isNavigatingToOnboardingFlow(action: NavigationAction): boolean {
```
Comment (linhvovan29546) [inline src/libs/Navigation/guards/OnboardingGuard.ts:137]: @MelvinBot Please use the same approach as isNavigatingToTestDriveModal

Comment (MelvinBot) [reply src/libs/Navigation/guards/OnboardingGuard.ts:137]: Refactored `isNavigatingToOnboardingFlow` to match the `isNavigatingToTestDriveModal` pattern:

1. Now accepts `state` parameter and checks the current focused route first (via `findFocusedRoute`)
2. Uses string literal `'RESET'` instead of `CONST.NAVIGATION_ACTIONS.RESET`
3. Uses `findFocusedRoute` on the action payload instead of manually iterating routes
4. Removed the extra `NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR` routes array check
5. Removed the now-unused `NAVIGATORS` import from both the guard and test files

All 17 tests pass.

Comment (MelvinBot) [reply src/libs/Navigation/guards/OnboardingGuard.ts:137]: Refactored `isNavigatingToOnboardingFlow` to match the `isNavigatingToTestDriveModal` pattern from `TestDriveModalGuard`:

1. Now takes `state` as a parameter and checks the current focused route first (via `findFocusedRoute(state)`)
2. Then checks the RESET action payload's focused route
3. Removed the extra `routes.some()` check for `NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR` — `findFocusedRoute` already traverses into nested states, so this is unnecessary

All 17 tests pass.

File: src/libs/Navigation/guards/OnboardingGuard.ts [hunk 2]
```diff
@@ -129,6 +129,26 @@ function shouldPreventReset(state: NavigationState, action: NavigationAction) {
...
+    if (action.type === CONST.NAVIGATION_ACTIONS.RESET && action.payload) {
+        const targetRoute = findFocusedRoute(action.payload as NavigationState);
+        if (isOnboardingFlowName(targetRoute?.name)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
```
Comment (linhvovan29546) [inline src/libs/Navigation/guards/OnboardingGuard.ts]:
```suggestion
function isNavigatingToOnboardingFlow(action: NavigationAction): boolean {
    // For NAVIGATE/PUSH actions, check if the action targets the OnboardingModalNavigator directly
    if (
        (action.type === CONST.NAVIGATION.ACTION_TYPE.NAVIGATE || action.type === CONST.NAVIGATION.ACTION_TYPE.PUSH) &&
        (action.payload as {name?: string} | undefined)?.name === NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR
    ) {
        return true;
    }

    return false;
}

```

**Comment by linhvovan29546** [comment]
@MelvinBot Apply this suggestion https://github.com/Expensify/App/pull/83193#discussion_r2840486912

File: src/libs/Navigation/guards/OnboardingGuard.ts (hunk 2 above)
Comment (MelvinBot) [reply src/libs/Navigation/guards/OnboardingGuard.ts]: Applied this suggestion. The isNavigatingToOnboardingFlow function now checks for NAVIGATE/PUSH actions targeting NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR instead of checking RESET action payloads with findFocusedRoute. Updated tests accordingly — all 18 tests pass.

File: src/libs/Navigation/guards/OnboardingGuard.ts [hunk 3]
```diff
@@ -129,6 +130,22 @@ function shouldPreventReset(state: NavigationState, action: NavigationAction) {
...
+/**
+ * Check if the navigation action is targeting an onboarding screen.
+ * This handles NAVIGATE/PUSH actions that target the OnboardingModalNavigator directly.
+ */
+function isNavigatingToOnboardingFlow(action: NavigationAction): boolean {
+    // For NAVIGATE/PUSH actions, check if the action targets the OnboardingModalNavigator directly
+    if (
+        (action.type === CONST.NAVIGATION.ACTION_TYPE.NAVIGATE || action.type === CONST.NAVIGATION.ACTION_TYPE.PUSH) &&
+        (action.payload as {name?: string} | undefined)?.name === NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR
+    ) {
```
[the remark this answers is not in the document]
Comment (linhvovan29546) [reply src/libs/Navigation/guards/OnboardingGuard.ts:141]: **AI GENERATED CONTENT:**

`linkTo` deliberately uses the default minimal `getStateFromPath` to avoid this — so clicking a Concierge link always produces `NAVIGATE` → `PUSH` (line 160 converts it). There is no `RESET` dispatched through the guard for in-app link clicks.

For a fresh page load to /onboarding/purpose (e.g. typing in address bar on a new tab), React Navigation sets the initial state via `getInitialState` and passes it directly as `initialState` to `NavigationContainer` — it doesn't dispatch a `RESET` action through the guard at all.

The only way a `RESET` reaches the guard would be from direct code calling `navigation.reset()`, which is not the scenario described in the issue.

**Review by linhvovan29546 — APPROVED** [review]
LGTM!

File: src/languages/de.ts [hunk 4]
```diff
@@ -65,8 +65,11 @@ import type {
     UpdatedPolicyBudgetNotificationParams,
     UpdatedPolicyCategoriesParams,
     UpdatedPolicyCategoryMaxAmountNoReceiptParams,
+    UpdatedPolicyCurrencyDefaultTaxParams,
+    UpdatedPolicyCustomTaxNameParams,
     UpdatedPolicyCustomUnitSubRateParams,
     UpdatedPolicyDefaultTitleParams,
+    UpdatedPolicyForeignCurrencyDefaultTaxParams,
```
Comment (situchan) [inline src/languages/de.ts:72]: @MelvinBot please pull main. The changes in this file is no longer needed.

File: src/libs/Navigation/guards/OnboardingGuard.ts [hunk 5]
```diff
@@ -129,6 +130,22 @@ function shouldPreventReset(state: NavigationState, action: NavigationAction) {
     return false;
 }
 
+/**
+ * Check if the navigation action is targeting an onboarding screen.
+ * This handles NAVIGATE/PUSH actions that target the OnboardingModalNavigator directly.
+ */
+function isNavigatingToOnboardingFlow(action: NavigationAction): boolean {
```
Comment (situchan) [inline src/libs/Navigation/guards/OnboardingGuard.ts:137]: @MelvinBot add unit test for this new function

File: src/languages/de.ts (hunk 4 above)
Comment (MelvinBot) [reply src/languages/de.ts:72]: Merged main into the branch. The changes to src/languages/de.ts are no longer in the diff — they were resolved by the merge. The PR now only contains changes to OnboardingGuard.ts and OnboardingGuard.test.ts.

File: src/libs/Navigation/guards/OnboardingGuard.ts (hunk 5 above)
Comment (MelvinBot) [reply src/libs/Navigation/guards/OnboardingGuard.ts:137]: @situchan

Comment (MelvinBot) [reply src/libs/Navigation/guards/OnboardingGuard.ts:137]: Added two additional edge case tests for the `isNavigatingToOnboardingFlow` function (20 tests total now pass):

1. **RESET with onboarding routes** — verifies RESET actions containing onboarding routes in the payload are NOT intercepted by the redirect (since the function only checks NAVIGATE/PUSH)
2. **REPLACE targeting onboarding** — verifies REPLACE actions targeting `ONBOARDING_MODAL_NAVIGATOR` are NOT intercepted

These complement the existing tests that cover the positive cases (NAVIGATE and PUSH to onboarding → redirect to HOME).

**Review by situchan — COMMENTED** [review]
@MelvinBot update Tests step 4 a bit.
`/onboarding/work` path not exist. It leads to not found page. The correct path is `/onboarding/work-email`

File: src/libs/Navigation/guards/OnboardingGuard.ts [hunk 6]
```diff
@@ -129,6 +130,22 @@ function shouldPreventReset(state: NavigationState, action: NavigationAction) {
...
+    // For NAVIGATE/PUSH actions, check if the action targets the OnboardingModalNavigator directly
+    if (
+        (action.type === CONST.NAVIGATION.ACTION_TYPE.NAVIGATE || action.type === CONST.NAVIGATION.ACTION_TYPE.PUSH) &&
+        (action.payload as {name?: string} | undefined)?.name === NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR
```
[the remark this answers is not in the document]
Comment (situchan) [reply src/libs/Navigation/guards/OnboardingGuard.ts:140]: See https://github.com/Expensify/App/pull/83193#discussion_r2840664422

File: src/components/ConnectToNetSuiteFlow/index.tsx [hunk 7]
```diff
@@ -8,11 +8,11 @@ import {isAuthenticationError} from '@libs/actions/connections';
 import {getAdminPoliciesConnectedToNetSuite} from '@libs/actions/Policy/Policy';
 import Navigation from '@libs/Navigation/Navigation';
 import {useAccountingContext} from '@pages/workspace/accounting/AccountingContext';
+import {getInitialSubPageForNetsuiteTokenInput} from '@pages/workspace/accounting/netsuite/utils';
```
Comment (situchan) [inline src/components/ConnectToNetSuiteFlow/index.tsx:11]: FYI: This prettier issue is being fixed in #83221

**Review by situchan — APPROVED** [review]

File: src/libs/Navigation/guards/OnboardingGuard.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (situchan) [reply src/libs/Navigation/guards/OnboardingGuard.ts:140]: Again, see https://github.com/Expensify/App/pull/83193#discussion_r2840664422

File: src/components/ConnectToNetSuiteFlow/index.tsx (hunk 7 above)
Comment (situchan) [reply src/components/ConnectToNetSuiteFlow/index.tsx:11]: It's merged now.
@MelvinBot pull main again

File: src/libs/Navigation/guards/OnboardingGuard.ts [hunk 8]
```diff
@@ -129,6 +130,22 @@ function shouldPreventReset(state: NavigationState, action: NavigationAction) {
     return false;
 }
 
+/**
+ * Check if the navigation action is targeting an onboarding screen.
+ * This handles NAVIGATE/PUSH actions that target the OnboardingModalNavigator directly.
+ */
+function isNavigatingToOnboardingFlow(action: NavigationAction): boolean {
+    // For NAVIGATE/PUSH actions, check if the action targets the OnboardingModalNavigator directly
```
Comment (Beamanator) [inline src/libs/Navigation/guards/OnboardingGuard.ts]: @MelvinBot this comment looks like a duplicate of the function description `This handles NAVIGATE/PUSH actions that target the OnboardingModalNavigator directly.` - keep one or the other, not both

Comment (MelvinBot) [reply src/libs/Navigation/guards/OnboardingGuard.ts]: Removed the duplicate inline comment and kept the JSDoc description above the function.

**Review by Beamanator — APPROVED** [review]
Loving the minimal changes with all the clarified & added tests! 🙏

**Comment by situchan** [comment]
> @MelvinBot update Tests step 4 a bit. `/onboarding/work` path not exist. It leads to not found page. The correct path is `/onboarding/work-email`

Melvin skipped [my message](https://github.com/Expensify/App/pull/83193#pullrequestreview-3841976063) 😄
